packages/parser/src/ethereum/index.ts

```diff
@@ -0,0 +1,180 @@
+import { BigNumber } from 'bignumber.js'
+import { ethers } from 'ethers'
+import { Tx } from '@shapeshiftoss/blockbook'
+import { Tx as ParseTx, TxSpecific as ParseTxSpecific, Token, TransferType, Transfer } from '../types'
+import { InternalTx, Network } from './types'
+import { getSigHash } from './utils'
+import * as multiSig from './multiSig'
+import * as thor from './thor'
+import * as uniV2 from './uniV2'
+import * as zrx from './zrx'
+
+export interface TransactionParserArgs {
+  network?: Network
+  midgardUrl: string
+  rpcUrl: string
+}
+
+export class TransactionParser {
+  network: Network
+
+  private thor: thor.Parser
+  private uniV2: uniV2.Parser
+  private zrx: zrx.Parser
+
+  constructor(args: TransactionParserArgs) {
+    const provider = new ethers.providers.JsonRpcProvider(args.rpcUrl)
+
+    this.network = args.network ?? Network.Mainnet
+
+    this.thor = new thor.Parser({ network: this.network, midgardUrl: args.midgardUrl })
+    this.uniV2 = new uniV2.Parser({ network: this.network, provider })
+    this.zrx = new zrx.Parser()
+  }
+
+  // return any addresses that can be detected
+  getInternalAddress(inputData: string): string | undefined {
+    switch (getSigHash(inputData)) {
+      case this.thor.transferOutSigHash:
+        return this.thor.getInternalAddress(inputData)
+      case multiSig.SENDMULTISIG_SIG_HASH:
+        return multiSig.getInternalAddress(inputData)
+      default:
+        return
+    }
+  }
+
+  async parse(tx: Tx, address: string, internalTxs?: Array<InternalTx>): Promise<ParseTx> {
+    const sendAddress = tx.vin[0].addresses?.[0] ?? ''
+    const receiveAddress = tx.vout[0].addresses?.[0] ?? ''
+
+    let result: ParseTxSpecific | undefined
+    switch (receiveAddress) {
+      case zrx.PROXY_CONTRACT: {
+        result = this.zrx.parse(tx)
+        break
+      }
+      case this.thor.routerContract: {
+        result = await this.thor.parse(tx)
+        break
+      }
+      case uniV2.ROUTER_CONTRACT: {
+        result = await this.uniV2.parse(tx)
+        break
+      }
+    }
```

Comment:
I would invert this logic and ask each parser if it wants to parse the tx, or you just call `parse` on all the available parsers (maybe stop at the first one that returns a result, or just accept all the results if we think that we want more than one parser to be able to process a transaction).

Reply:
I like this concept. Another aspect for the continued improvement I will add as an issue to discuss and implement after initial cut.